pkg/util/loop/loop.go

```diff
@@ -253,39 +220,157 @@ func (loop *Device) attachLoop(image *os.File, mode int, number *int) error {
+		// may be slow, so setting a time interval and number of retries may be hazardous, and trying other loop devices
+		// just deport the issue to the next devices as falsely stated here https://dev.arvados.org/issues/18489.
+		// So retry 5 times with a sleep period of 200ms between each attempt.
+		if retryCount < sleepRetries {
+			time.Sleep(sleepInterval)
+			return errStatusTryAgain
+		} else if retryCount == sleepRetries {
+			// The sleeping period is over and there is remaining dirty pages in cache for the corresponding image,
+			// let's use the rough approach to flush cached pages to filesystem, the kernel is not providing a way to
+			// flush and wait, syncfs/fsync/sync_file_range are not working as expected here, so we call flushCache
+			// which will try to issue a block device flush command when the image is located on a block device, if the
+			// image is on a shared storage, a ramfs or anything else which isn't a block device, a sync syscall is
+			// issued with the costs it involved.
+			//
+			// Dear reader, if you are not satisfied by the approach, you are invited to reproduce the issue first by using
+			// an Ubuntu 18.04 distribution containing the fix/bug above and build a small image like:
+			//
+			// $ apptainer build /tmp/busy.sif docker://busybox
+			// $ for i in $(seq 1 100); do apptainer exec /tmp/busy.sif true; done
+			//
+			// And search a more elegant solution
+			if err := flushCache(imageFd, imageInfo); err != nil {
+				return fmt.Errorf("while syncing/flushing image cache: %s", err)
+			}
+			return errStatusTryAgain
+		} else if retryCount == sleepRetries+1 {
+			// e2e tests have shown that the sync approach is not sufficient under high load
+			// circumstances, therefore we are giving an additional grace period, after that
+			// we are over and return a cache invalidate too slow error
+			time.Sleep(flushGracePeriod)
+			return errStatusTryAgain
+		}
+
+		return fmt.Errorf("failed to set loop device status (%s): cache invalidate too slow", loopDevice)
+	}
+}
+
+func flushCache(imageFd uintptr, imageInfo *syscall.Stat_t) error {
+	devStr := fmt.Sprintf("%d:%d", unix.Major(imageInfo.Dev), unix.Minor(imageInfo.Dev))
+	entries, err := proc.GetMountInfoEntry("/proc/self/mountinfo")
+	if err != nil {
+		return fmt.Errorf("while getting mountinfo: %s", err)
+	}
+	for _, e := range entries {
+		if e.Dev != devStr {
+			continue
+		}
+		fi, err := os.Stat(e.Source)
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+			return fmt.Errorf("while getting information for %s: %s", e.Source, err)
+		}
+		// not a block device
+		if fi.Mode()&os.ModeDevice == 0 {
+			continue
+		}
+		// trigger block device flush command
+		f, err := os.Open(e.Source)
+		if err != nil {
+			return fmt.Errorf("while opening %s: %s", e.Source, err)
+		}
+		defer f.Close()
+
+		_, _, esys := syscall.Syscall(syscall.SYS_IOCTL, f.Fd(), unix.BLKFLSBUF, 0)
```

Comment:
@cclerget something to consider ... bd_flush may not flush at the FS level, nor stop FS I/o in flight.   Noting that loop device is rare use-case where kernel internally actually captures an FD.    Seems like there should be some ioctl at the mount point which performs fs.flush properly

Reply:
I didn't find more granular than this call available from user space, this is called after the sleep period of 1s, so we are already in a worst case scenario, this is not 100% reliable and it's provided as best-effort, but as 5.8 kernel introduced `LOOP_CONFIGURE` ioctl allowing to set the file descriptor and the offset/sizelimit in a single ioctl command, this issue would disappear. I won't push other effort on fixing this one, I restored the old sleeping behavior and added some flushing helper, I'm done